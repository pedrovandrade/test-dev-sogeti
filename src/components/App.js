import React, { useState } from 'react';
// import TodoList from 'components/TodoList.js';
import TodoList from './TodoList';

/**
 * Generates a random UUID, either with the new crypto.randomUUID native method
 * or with a (less robust) Math.random solution
 */
function generateUUID() {
  try {
    return crypto.randomUUID();
  } catch { // In case crypto.randomUUID is not found
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
    });
  }
}

export default function App(props) {
  /** ******** REACT HOOKS ********* */
  const [newTodo, setNewTodo] = useState('');

  function addTodo() {
    if (newTodo === '') return; // Don't insert an empty string

    let newUuid = generateUUID();
    let alreadyExists = !!props.listData.find(item => item.uuid === newUuid);
    while (alreadyExists) {
      newUuid = generateUUID();
      alreadyExists = !!props.listData.find(item => item.uuid === newUuid);
    }

    props.setListData(oldArray => [{
      "title": newTodo,
      "state": "todo",
      "description": "",
      "uuid": newUuid
    }, ...oldArray]);

    setNewTodo('');
  }

  return (
    <div className='list-container' >
      <h1 className='list-title' >My todo list</h1>
      <
        input
        type='text'
        placeholder='What do you want to do?'
        value={newTodo}
        onInput={(e) => setNewTodo(e.target.value)}
      />
      <button id='add-button' onClick={addTodo}>Add</button>
      <TodoList listData={props.listData || []} setListData={props.setListData} />
    </div>
  );
}
