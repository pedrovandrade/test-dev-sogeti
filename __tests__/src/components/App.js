import React from 'react';
import renderer from 'react-test-renderer';
import App from '../../../src/components/App';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, getByPlaceholderText } from '@testing-library/react';

describe('src/components/App.js', () => {
  const mockList = [
    {
      title: 'Do groceries',
      state: 'todo',
      description: 'Go to the supermarket and buy eggs, butter, sugar, fruits, soap and toillete paper',
      uuid: '8348db18-83ba-11ec-a8a3-0242ac120002'
    },
    {
      title: 'Wash car',
      state: 'todo',
      description: 'Take the car to the gas station and leave it there for 30 minutes in the express washing',
      uuid: '9c377382-83ba-11ec-a8a3-0242ac120002'
    },
    {
      title: 'Clean house',
      state: 'todo',
      description: 'Pass vaccum, clean the surfaces, clean the toillete and pass water on the floor',
      uuid: 'a081b5ec-83ba-11ec-a8a3-0242ac120002'
    }
  ];

  it('should expose a function component', () => {
    expect(App).toBeInstanceOf(Function);
  });
  it('should render correctly', () => {
    const tree = renderer
    .create(
      <MemoryRouter>
        <App listData={mockList} />
      </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
  describe('When the app start', () => {
    it('should have an empty text input for inserting new todos', () => {
      const { container } = render(
        <MemoryRouter>
          <App listData={mockList} />
        </MemoryRouter>
      );
      const textInput = getByPlaceholderText(container, 'What do you want to do?');
      expect(textInput.textContent).toBe('');
    })
  });
});
