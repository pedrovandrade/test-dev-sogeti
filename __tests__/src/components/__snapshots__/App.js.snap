// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/App.js should render correctly 1`] = `
<div
  className="list-container"
>
  <h1
    className="list-title"
  >
    My todo list
  </h1>
  <input
    onInput={[Function]}
    placeholder="What do you want to do?"
    type="text"
    value=""
  />
  <button
    id="add-button"
    onClick={[Function]}
  >
    Add
  </button>
  <ul>
    <li>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <a
          href="/details/8348db18-83ba-11ec-a8a3-0242ac120002"
          onClick={[Function]}
          style={
            Object {
              "color": "black",
              "textDecoration": "none",
            }
          }
        >
          <div
            className="todo-item"
          >
            Do groceries
          </div>
        </a>
        <button
          className="delete-button"
          onClick={[Function]}
          type="button"
        />
      </label>
    </li>
    <li>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <a
          href="/details/9c377382-83ba-11ec-a8a3-0242ac120002"
          onClick={[Function]}
          style={
            Object {
              "color": "black",
              "textDecoration": "none",
            }
          }
        >
          <div
            className="todo-item"
          >
            Wash car
          </div>
        </a>
        <button
          className="delete-button"
          onClick={[Function]}
          type="button"
        />
      </label>
    </li>
    <li>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <a
          href="/details/a081b5ec-83ba-11ec-a8a3-0242ac120002"
          onClick={[Function]}
          style={
            Object {
              "color": "black",
              "textDecoration": "none",
            }
          }
        >
          <div
            className="todo-item"
          >
            Clean house
          </div>
        </a>
        <button
          className="delete-button"
          onClick={[Function]}
          type="button"
        />
      </label>
    </li>
  </ul>
</div>
`;
